### ISSUE TITLE
Adding query to some URLs creates incorrect paths

### ISSUE THREAD

**Issue by braco** [rackt/react-router#1148]
```
transitionTo/makePath('/foo/?bar=1', null, {anything: 'here'})
```
incorrectly produces
```
/foo/bar=1?anything=here
```
whereas
```
transitionTo/makePath('/foo/baz?bar=1', null, {anything: 'here'})
```
will create the correct URL
```
/foo/baz?bar=1&anything=here
```
note the incorrect URL path ends with /?, whereas the other has a final object: /baz?

**Comment by braco** [comment]
Additionally, even
```
transitionTo('/foo/?bar') 
```
with no extra params seems to remove the ?

**Comment by taurose** [comment]
You're not supposed to add query in `to` parameter, just the pathname (or route name). Use the third parameter as you've already shown.

```js
transitionTo/makePath('/foo/', null, {bar: 1, anything: 'here'})
```

**Comment by braco** [comment]
Then the behavior shouldn't change so drastically based on the URL fragments. It mangles the URL only when it ends with a trailing slash.

but either way, the latter issue still remains https://github.com/rackt/react-router/issues/1148#issuecomment-98020466

**Comment by appsforartists** [comment]
#1192 was a variant of this, but has been merged into this issue.

In that issue, I deduced that the problem appears to be this line:
```javascript
.replace(paramInjectTrailingSlashMatcher, '/');
```

**Comment by ryanflorence** [comment]
the new `history` lib that router is built on fixes this, as seen in rackt/history@d6380e5d
